File: src/main/java/it/scoppelletti/spaceship/gradle/android/tasks/CreditsTask.java
/*
 * Copyright (C) 2020 Dario Scoppelletti, <http://www.scoppelletti.it/>.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package it.scoppelletti.spaceship.gradle.android.tasks;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.io.Writer;
import java.net.URI;
import java.util.Objects;
import java.util.Set;
import java.util.TreeSet;
import javax.inject.Inject;
import it.scoppelletti.spaceship.gradle.android.AndroidLibraryTools;
import it.scoppelletti.spaceship.gradle.android.model.CreditItem;
import it.scoppelletti.spaceship.gradle.model.SpaceshipExtension;
import lombok.Getter;
import lombok.Setter;
import org.apache.commons.lang3.StringUtils;
import org.gradle.api.DefaultTask;
import org.gradle.api.artifacts.Configuration;
import org.gradle.api.artifacts.ExternalDependency;
import org.gradle.api.tasks.Input;
import org.gradle.api.tasks.OutputFile;
import org.gradle.api.tasks.TaskAction;
import org.gradle.tooling.BuildException;
import org.thymeleaf.ITemplateEngine;
import org.thymeleaf.context.Context;

/**
 * Generate the credit file.
 */
public class CreditsTask extends DefaultTask {

    /**
     * Namespace of the credit database XML file.
     */
    @SuppressWarnings("unused")
    public static final String NAMESPACE =
            "http://www.scoppelletti.it/ns/credits/1";

    /**
     * Template variable {@code credits}.
     */
    public static final String VAR_CREDITS = "credits";

    /**
     * Variant name.
     */
    @Getter(onMethod_ = { @Input })
    @Setter
    private String variantName;

    /**
     * Credit database URL.
     */
    @Getter(onMethod_ = { @Input })
    @Setter
    private URI databaseUrl;

    /**
     * Template name.
     */
    @Getter(onMethod_ = { @Input })
    @Setter
    private String templateName;

    /**
     * Credit output file
     */
    @Getter(onMethod_ = { @OutputFile })
    @Setter
    private File outputFile;

    /**
     * Sole constructor.
     */
    @Inject
    public CreditsTask() {
    }

    /**
     * Executes the task.
     */
    @TaskAction
    public void run() {
        String configName;
        File outDir;
        Configuration config;
        CreditDatabase database;
        SpaceshipExtension spaceshipExt;
        ITemplateEngine templEngine;
        Context templCtx;
        Set<CreditItem> credits;

        if (StringUtils.isBlank(variantName)) {
            throw new NullPointerException("Property variantName is null.");
        }

        Objects.requireNonNull(databaseUrl, "Property databaseUrl is null.");

        if (StringUtils.isBlank(templateName)) {
            throw new NullPointerException("Property templateName is null.");
        }

        Objects.requireNonNull(outputFile, "Property creditFile is null.");

        spaceshipExt = Objects.requireNonNull(getProject().getExtensions()
                .findByType(SpaceshipExtension.class), () ->
                String.format("Extension %1$s not found.",
                        SpaceshipExtension.class));

        database = CreditDatabase.load(databaseUrl);

        configName = variantName.concat(AndroidLibraryTools.CONFIG_RUNTIME);
        config = getProject().getConfigurations().getByName(configName);

        outDir = outputFile.getParentFile();
        if (outDir != null && !outDir.exists()) {
            //noinspection ResultOfMethodCallIgnored
            outDir.mkdirs();
        }

        credits = new TreeSet<>();
        database.getCredits().forEach(item -> {
            if (item.isForce()) {
                credits.add(item);
            }
        });

        config.getAllDependencies().withType(ExternalDependency.class)
                .forEach(dep -> addDependency(credits, dep, database));

        templCtx = new Context();
        templCtx.setVariable(CreditsTask.VAR_CREDITS, credits);
        templEngine = spaceshipExt.getTemplateEngine();

        try (Writer out = new FileWriter(outputFile);
             BufferedWriter writer = new BufferedWriter(out)) {
            templEngine.process(templateName, templCtx, writer);
        } catch (IOException ex) {
            throw new BuildException(ex.getMessage(), ex);
        }
    }

    private void addDependency(Set<CreditItem> credits, ExternalDependency dep,
            CreditDatabase database) {
        String artifactId, groupId;
        CreditItem credit;
        ArtifactItem artifact;

        groupId = Objects.requireNonNull(dep.getGroup(),
                "Argument dep.group is null.");
        artifactId = dep.getName();
        artifact = new ArtifactItem(groupId, artifactId);
        getLogger().debug("Detect artifact {}.", artifact);

        credit = database.getCreditByArtifact(artifact);
        if (credit == null) {
            getLogger().warn("No credit found for artifact {}.", artifact);
            return;
        }

        credits.add(credit);
    }
}

File: src/main/kotlin/it/scoppelletti/spaceship/gradle/android/tasks/CreditsTask.kt
/*
 * Copyright (C) 2020 Dario Scoppelletti, <http://www.scoppelletti.it/>.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package it.scoppelletti.spaceship.gradle.android.tasks

import freemarker.template.TemplateException
import it.scoppelletti.spaceship.gradle.android.model.ArtifactElement
import it.scoppelletti.spaceship.gradle.android.model.CreditElement
import it.scoppelletti.spaceship.gradle.android.model.CreditsModel
import it.scoppelletti.spaceship.gradle.freemarker.FreemarkerExt
import java.io.IOException
import javax.inject.Inject
import org.gradle.api.DefaultTask
import org.gradle.api.artifacts.ExternalDependency
import org.gradle.api.file.DirectoryProperty
import org.gradle.api.model.ObjectFactory
import org.gradle.api.provider.Property
import org.gradle.api.provider.ProviderFactory
import org.gradle.api.tasks.Input
import org.gradle.api.tasks.OutputDirectory
import org.gradle.api.tasks.TaskAction
import org.gradle.tooling.BuildException

private const val CONFIG_RUNTIME = "RuntimeElements"
private const val TEMPLATE_NAME =
    "/it/scoppelletti/spaceship/gradle/android/tasks/credits.ftl"

/**
 * Generates the credits file.
 *
 * @since 1.0.0
 *
 * @property variantName  Variant name.
 * @property databaseUrl  Credits database URL.
 * @property templateName Template name.
 * @property outputDir    Output folder.
 * @property creditsName  Name of the credits file.
 */
public abstract class CreditsTask @Inject constructor(
    objects: ObjectFactory,
    providers: ProviderFactory
): DefaultTask() {

    @get:Input
    public abstract val variantName: Property<String>

    @get:Input
    public val databaseUrl: Property<String> =
        objects.property(String::class.java).apply {
            convention(providers.gradleProperty(PROP_DATABASE))
        }

    @get:Input
    public val templateName: Property<String> =
        objects.property(String::class.java).apply {
            convention(providers.gradleProperty(PROP_TEMPLATE)
                .orElse(TEMPLATE_NAME))
        }

    @get:OutputDirectory
    public abstract val outputDir: DirectoryProperty

    @get:Input
    public val creditsName: Property<String> =
        objects.property(String::class.java).apply {
            convention(CREDITS_NAME)
        }

    @TaskAction
    public fun action() {
        val templName = templateName.get()
        val outDir = outputDir.get().asFile
        val fileName = creditsName.get()
        val database = CreditDatabase.load(databaseUrl.get())
        val configName = variantName.get() + CONFIG_RUNTIME
        val config = project.configurations.getByName(configName)

        val credits = mutableSetOf<CreditElement>()
        database.credits
            .filter { it.force }
            .toCollection(credits)

        config.allDependencies.withType(ExternalDependency::class.java)
            .forEach {
                addDependency(credits, it, database)
            }

        val model = CreditsModel(credits.sortedBy { it.key })

        val templ = try {
            FreemarkerExt.config.getTemplate(templName)
        } catch (ex: IOException) {
            logger.error("Fail to get $templName.", ex)
            throw BuildException("Fail to get $templName.", ex)
        }

        outDir.mkdirs()
        outDir.resolve(fileName).writer().use { writer ->
            try {
                templ.process(model, writer)
            } catch (ex: IOException) {
                logger.error("Fail to process $templName.", ex)
                throw BuildException("Fail to process $templName.", ex)
            } catch (ex: TemplateException) {
                logger.error("Fail to process $templName.", ex)
                throw BuildException("Fail to process $templName.", ex)
            }
        }
    }

    private fun addDependency(credits: MutableSet<CreditElement>,
        dep: ExternalDependency, database: CreditDatabase) {
        val artifact = ArtifactElement(
            requireNotNull(dep.group) {
                "Argument dep.group is null."
            },
            dep.name
        )
        logger.debug("Detect artifact {}", artifact)

        database[artifact]?.let {
            credits.add(it)
        } ?: run {
            logger.warn("No credit found for artifact {}.", artifact)
        }
    }

    public companion object {

        /**
         * Name of the credits file.
         */
        public const val CREDITS_NAME: String = "credits.html"

        /**
         * Property containing the URL of the credits database.
         */
        public const val PROP_DATABASE: String =
            "it.scoppelletti.spaceship.credits.databaseUrl"

        /**
         * Property containing the URL of the credits template.
         */
        public const val PROP_TEMPLATE: String =
            "it.scoppelletti.spaceship.credits.templateName"

        /**
         * Base of the name of the schema.
         */
        public const val SCHEMA: String =
            "spaceship/gradle/android/tasks/credits"

        /**
         * Version of the schema.
         */
        public const val VERSION: Int = 1
    }
}